source/base/function_lib.cc

```diff
@@ -2112,8 +2112,12 @@ namespace Functions
 
     double prod = 1;
     for (unsigned int s=0; s<dim; ++s)
-      prod *= std::pow(p[s], exponents[s]);
-
+      {
+        if (p[s] < 0)
+          Assert(std::floor(exponents[s]) == exponents[s],
+                 ExcInternalError());
```

Comment:
I think ExcInternalError is the wrong category here. This is a completely reasonable error message that does not indicate that some internal data structures went out of whack or the algorithm was wrongly implemented.

Can you use ExcMessage with a descriptive text for what it is that's going wrong?

(Same in the corresponding place below.)